File: Plugins/iOS/RudderPlugin_iOS/RudderPlugin_iOS/Source/RudderElementBuilder.swift
    //
//  RudderElementBuilder.swift
//  RudderPlugin_iOS
//
//  Created by Arnab Pal on 14/09/19.
//  Copyright © 2019 Arnab Pal. All rights reserved.
//

import Foundation

class RudderElementBuilder : NSObject {
    private var eventName: String? = nil
    
    @objc func withEventName(eventName: String) -> RudderElementBuilder {
        self.eventName = eventName
        return self
    }
    
    private var userId: String? = nil
    
    @objc func withUserId(userId: String) -> RudderElementBuilder {
        self.userId = userId
        return self
    }
    
    private var eventProperties: Dictionary<String, NSObject>? = nil
    @objc func withEventProperties(properties: Dictionary<String, NSObject>) -> RudderElementBuilder {
        self.eventProperties = properties
        return self
    }
    
    @objc func withEventProperty(key: String, value: NSObject) -> RudderElementBuilder {
        if(self.eventProperties == nil) {
            self.eventProperties = Dictionary()
        }
        self.eventProperties![key] = value
        return self
    }
    
    private var userProperties: Dictionary<String, NSObject>? = nil
    @objc func withUserProperties(properties: Dictionary<String, NSObject>) -> RudderElementBuilder {
        self.userProperties = properties
        return self
    }
    
    @objc func withUserProperty(key: String, value: NSObject) -> RudderElementBuilder {
        if (self.userProperties == nil) {
            self.userProperties = Dictionary()
        }
        self.userProperties![key] = value
        return self
    }
    
    private var integrations: Dictionary<String, Bool>? = nil
    @objc func withIntegration(integrations: Dictionary<String, Bool>) -> RudderElementBuilder {
        self.integrations = integrations
        return self
    }
    
    @objc func withIntegration(integration: String, isEnabled: Bool) -> RudderElementBuilder {
        if (self.integrations == nil) {
            self.integrations = Dictionary()
        }
        self.integrations![integration] = isEnabled
        return self
    }
    
    @objc func build() -> RudderElement {
        let element = RudderElement()
        if (self.eventName != nil) {
            element.setEventName(eventName: self.eventName!)
        }
        if (self.userId != nil) {
            element.setUserId(userId: self.userId!)
        }
        if (self.eventProperties != nil) {
            element.setEventProperties(eventProperties: AnyCodable(value: self.eventProperties!))
        }
        if (self.userProperties != nil) {
            element.setUserProperties(userProperties: AnyCodable(value: self.userProperties!))
        }
        if (self.integrations != nil) {
            element.setIntegrations(integrations: AnyCodable(value: self.integrations!))
        }
        return element
    }
}

File: RudderUnityPluginDemo/Assets/RudderUnityPlugin/RudderElementBuilder.cs
using System.Collections.Generic;

public class RudderElementBuilder
{
    private string eventName;
    public RudderElementBuilder WithEventName(string eventName)
    {
        this.eventName = eventName;
        return this;
    }
    private string userId;
    public RudderElementBuilder WithUserId(string userId)
    {
        this.userId = userId;
        return this;
    }
    private Dictionary<string, object> eventProperties;
    public RudderElementBuilder WithEventProperties(Dictionary<string, object> eventProperties)
    {
        if (eventProperties == null)
        {
            // do not set null value
            return this;
        }
        else
        {
            if (this.eventProperties == null)
            {
                this.eventProperties = new Dictionary<string, object>();
            }
            foreach (var key in eventProperties.Keys)
            {
                var value = eventProperties[key];
                if (value != null)
                {
                    this.eventProperties.Add(key, value);
                }
            }
        }
        return this;
    }
    public RudderElementBuilder WithEventProperty(string key, object value)
    {
        if (this.eventProperties == null)
        {
            this.eventProperties = new Dictionary<string, object>();
        }
        if (value != null)
        {
            this.eventProperties.Add(key, value);
        }
        return this;
    }
    private Dictionary<string, object> userProperties;
    public RudderElementBuilder WithUserProperties(Dictionary<string, object> userProperties)
    {
        if (userProperties == null)
        {
            // do not set null values
            return this;
        }
        else
        {
            if (this.userProperties == null)
            {
                this.userProperties = new Dictionary<string, object>();
            }
            foreach (var key in userProperties.Keys)
            {
                var value = userProperties[key];
                if (value != null)
                {
                    this.userProperties.Add(key, value);
                }
            }
        }
        return this;
    }
    public RudderElementBuilder WithUserProperty(string key, object value)
    {
        if (this.userProperties == null)
        {
            this.userProperties = new Dictionary<string, object>();
        }
        if (value != null)
        {
            this.userProperties.Add(key, value);
        }
        return this;
    }
    public RudderElement Build()
    {
        RudderElement element = new RudderElement();
        element.eventName = this.eventName;
        element.userId = this.userId;
        element.eventProperties = this.eventProperties;
        element.userProperties = this.userProperties;
        return element;
    }

}